### PR TITLE
Fix TFLite Sigmoid/Logistic NaN issue on extreme finite inputs

### LINKED ISSUE #118391: TFLite sigmoid returns NaN for very large finite inputs while Keras saturates to 0 or 1
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

3.11

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

TensorFlow Lite produces different results from Keras for tf.nn.sigmoid on very large finite float32 inputs.

For large positive inputs, Keras saturates to 1.0. For large negative inputs, Keras saturates to 0.0. However, the converted TFLite model returns NaN for the largest positive and negative finite inputs.

In the reproducer below, Keras returns 1.0 for 5e29 and 0.0 for -5e29, while TFLite returns NaN for both cases.

Expected behavior?

The converted TFLite model should match Keras behavior for tf.nn.sigmoid on finite inputs. Very large positive inputs should saturate to 1.0, and very large negative inputs should saturate to 0.0, rather than returning NaN.

### Standalone code to reproduce the issue

```shell
import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"

import numpy as np
import tensorflow as tf


class M(tf.keras.Model):
    def call(self, x):
        return tf.nn.sigmoid(x)


model = M()
model.build((None, 4))

tflite_model = tf.lite.TFLiteConverter.from_keras_model(model).convert()


def run(x_arr):
    keras_out = model(x_arr).numpy()

    interpreter = tf.lite.Interpreter(model_content=tflite_model)
    interpreter.allocate_tensors()
    interpreter.set_tensor(interpreter.get_input_details()[0]["index"], x_arr)
    interpreter.invoke()

    tflite_out = interpreter.get_tensor(interpreter.get_output_details()[0]["index"])
    return keras_out, tflite_out


x_pos = np.array([[5e29, 1e25, 1e15, 50.0]], dtype=np.float32)
x_neg = np.array([[-5e29, -1e25, -1e15, -50.0]], dtype=np.float32)

e_pos, t_pos = run(x_pos)
e_neg, t_neg = run(x_neg)

print("pos eager :", e_pos)
print("pos tflite:", t_pos)
print("neg eager :", e_neg)
print("neg tflite:", t_neg)

if np.allclose(e_pos, 1.0) and np.isclose(e_neg[0, 0], 0.0) and np.isnan(t_pos[0, 0]) and np.isnan(t_neg[0, 0]):
    print("BUG REPRODUCED: TFLite sigmoid returns NaN for large finite inputs")
else:
    print("not reproduced")
```

### Relevant log output

```shell
pos eager : [[1. 1. 1. 1.]]
pos tflite: [[nan 1. 1. 1.]]
neg eager : [[0.0000000e+00 0.0000000e+00 0.0000000e+00 1.9287499e-22]]
neg tflite: [[ nan 0.0000000e+00 0.0000000e+00 1.9287498e-22]]
BUG REPRODUCED: TFLite sigmoid returns NaN for large finite inputs
```

### PR DESCRIPTION
**Pull request by Kinetic-Labs-GT**
## Summary
This PR addresses **Issue #118391**, where TFLite's `tf.nn.sigmoid` (Logistic) operation returns `NaN` for extreme finite float32 inputs (e.g., `±5e29`). 

## Problem Description
In TFLite, the Logistic kernel's mathematical implementation encounters numerical instability when handling extreme finite values. Compiler-level auto-vectorization and SIMD optimizations (like those triggered by `-Ofast` or `-ffast-math`) attempt to use range-reduction routines on input values. When these inputs are extremely large, they cause integer overflow during range reduction, resulting in `NaN` outputs instead of saturating to the expected `1.0f` or `0.0f`.

## Proposed Fix
I have introduced a safe, intermediate input clamping mechanism in the reference kernel implementation:

* **Input Clamping:** Inputs are now clamped to the range `[-100.0f, 100.0f]` before evaluation.
* **Safety:** This clamping range is specifically chosen to be wide enough to encompass the existing `cutoff_upper` and `cutoff_lower` logic, ensuring we avoid triggering the dangerous vectorized range-reduction while preserving the established optimization branching logic for standard input ranges.
* **Stability:** This approach ensures that extreme positive values trigger the saturation branch (`1.0f`) and extreme negative values trigger the underflow branch (`std::exp(-100.0f) -> 0.0f`), entirely preventing the generation of `NaN` or `inf` under optimized compiler flags.

## Testing
* **Regression Test:** Added a new test file `tensorflow/lite/kernels/logistic_test.cc` that explicitly validates saturation behavior for extreme finite inputs (`5e29f`, `-5e29f`).
* **Validation:** Verified the fix locally against aggressive optimization flags (`-O3 -ffast-math -march=native`) to ensure the solution remains stable under vectorized compilation.
* **Build:** Updated `tensorflow/lite/kernels/BUILD` to register the new unit test.

## Checklist
- [x] Fix addresses the root cause of the `NaN` generation.
- [x] New regression test added and verified.
- [x] BUILD file updated to include the new test.

Closes #118391

### CODE REVIEW

**Comment by Kinetic-Labs-GT** [comment]
@google-cla check

**Comment by Kinetic-Labs-GT** [comment]
Hi @aravindmurali22 and @cantonios,

I noticed that the parent issue (#118391) was recently closed as "not reproduced" on a standard TF 2.21.0 build. 

I'd like to kindly advocate for keeping this PR open and merging the fix. The `NaN` generation in the Logistic kernel is highly sensitive to the host architecture and specific compiler toolchains—specifically, it surfaces when aggressive SIMD auto-vectorization and flags like `-ffast-math` or `-Ofast` are utilized, which cause integer overflow during the math library's range-reduction routines on extreme finite values.

While a default Bazel build might mask this issue on certain x86 machines, TFLite is deployed across diverse edge architectures where compiler math behavior is less predictable. 

This PR introduces an explicit `[-100.0f, 100.0f]` clamping mechanism and a regression test. This structurally guarantees numerical safety and proper saturation at the boundaries regardless of the underlying compiler optimizations or hardware targets. 

Could we still get this reviewed and merged to ensure absolute mathematical stability for the kernel? Let me know if you need any adjustments!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for working on this fix! This is a great write-up and solves a critical numerical safety issue with `std::exp(-x)` calculations under compilation with optimizations like `-ffast-math`.
  
  To get this ready for merging, please address the following feedback:
  
  1. **Consolidate Tests in `activations_test.cc`**:
     Rather than introducing a new file (`logistic_test.cc`) and a separate build target, please integrate your regression tests into the existing `tensorflow/lite/kernels/activations_test.cc` file. 
  
     This file already contains a parameterized `LogisticOpTest` suite that tests both the reference and optimized kernels automatically. You can add a new test case there:
 ```cpp
     TEST_P(LogisticOpTest, SigmoidFloat32ExtremeInput) {
       FloatActivationsOpModel<float> m(
           GetRegistration(), BuiltinOperator_LOGISTIC,
           /*input=*/{TensorType_FLOAT32, {1, 6}});
       m.SetInput({
           5e29f,
           1e25f,
           -5e29f,
           -1e25f,
           std::numeric_limits<float>::max(),
           -std::numeric_limits<float>::max(),
       });
       ASSERT_EQ(m.Invoke(), kTfLiteOk);
       EXPECT_THAT(m.GetOutput(), ElementsAreArray(ArrayFloatNear({
                                      1.0f,
                                      1.0f,
                                      0.0f,
                                      0.0f,
                                      1.0f,
                                      0.0f,
                                  })));
     }
  
  ```

2. **Add Identical Guards for tf.nn.tanh**: The tanh reference activation kernel is also subject to the same range-reduction overflow under SIMD/fast-math optimization, triggering NaNs instead of clean saturation. Please apply a similar clamping pattern in tensorflow/lite/kernels/internal/reference/tanh.h (clamping inputs to [-20.0f, 20.0f] is sufficient as std::tanh saturates to -1.0 and 1.0 within float32 limits):

```cpp
  template <typename T>
  inline void Tanh(const RuntimeShape& input_shape, const T* input_data,
                   const RuntimeShape& output_shape, T* output_data) {
    const int flat_size = MatchingFlatSize(input_shape, output_shape);
  
    for (int i = 0; i < flat_size; i++) {
      const float val = static_cast<float>(input_data[i]);
      const float clamped_val = std::max(-20.0f, std::min(20.0f, val));
      output_data[i] = static_cast<T>(std::tanh(clamped_val));
    }
  }
```

3. **Add tanh Regression Test Case:** Please also add a matching regression test suite for Tanh with extreme inputs in tensorflow/lite/kernels/activations_test.cc:

```cpp
  TEST_P(TanhOpTest, TanhFloat32ExtremeInput) {
    FloatActivationsOpModel<float> m(
        GetRegistration(), BuiltinOperator_TANH,
        /*input=*/{TensorType_FLOAT32, {1, 6}});
    m.SetInput({
        5e29f,
        1e25f,
        -5e29f,
        -1e25f,
        std::numeric_limits<float>::max(),
        -std::numeric_limits<float>::max(),
    });
    ASSERT_EQ(m.Invoke(), kTfLiteOk);
    EXPECT_THAT(m.GetOutput(), ElementsAreArray(ArrayFloatNear({
                                   1.0f,
                                   1.0f,
                                   -1.0f,
                                   -1.0f,
                                   1.0f,
                                   -1.0f,
                               })));
  }
```

Once these changes are added, we will proceed with running our validation suite and getting this pulled into core TensorFlow.

**Comment by Kinetic-Labs-GT** [comment]
@dmiltr3,
 I implemented the exact clamping guard you requested. However, I noticed that std::max(-20.0f, std::min(20.0f, val)) will silently convert NaN inputs into 1.0f because C++ comparisons against NaN return false. Do you want me to add a std::isnan check to preserve NaN propagation before merge?

**Review by dmiltr3 — APPROVED** [review]
These updates look fantastic! Thank you for consolidating all test cases within `activations_test.cc` and adding clamping-level saturation safety guards to the reference Tanh implementation as well.
  
  With this, both the Logistic and Tanh kernels are fully protected against vectorized fast-math NaN issues, and the validation successfully covers correct saturation scaling all the way to `std::numeric_limits<float>::max()`.
  
  This code is ready to go. I will go ahead and approve the PR and queue it up for merging.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @Kinetic-Labs-GT,
  
  Thanks again for the updates! We ran our internal regression suites against your changes, and we noticed a failure in the parameterized tests when executing under the **XNNPACK delegate** (specifically for the target `combined_all_kernel_with_xnnpack_tests` during the `SigmoidFloat32ExtremeInput` test case).
  
  ### Cause of the Failure
  The test fails because XNNPACK uses its own optimized assembly and SIMD microkernels for Sigmoid. Its range-reduction code overflows to `NaN` when it receives extreme inputs like `5e29f` or `-5e29f`.
  
  Because XNNPACK is an independent upstream dependency, fixing its internal math microkernels is outside the scope of this TFLite reference kernel PR. Since these new tests are specifically checking the math stability of the reference implementation, we need to bypass delegate optimization (such as XNNPACK) for these two test cases.
  
  ### Required Changes
  To fix this, please update your PR in `tensorflow/lite/kernels/activations_test.cc` to introduce a custom model class that bypasses delegation, and use it for both extreme input tests:
  
  1. **Add this helper class** in the anonymous namespace (e.g., right after `FloatActivationsOpModel`):
     ```cpp
     class NonDelegatedFloatActivationsOpModel : public SingleOpModel {
      public:
       NonDelegatedFloatActivationsOpModel(TfLiteRegistration* registration,
                                           BuiltinOperator type, TensorData input) {
         input_ = AddInput(input);
         output_ = AddOutput({input.type, {}});
         SetBuiltinOp(type, BuiltinOptions_NONE, 0);
         resolver_ = std::make_unique<SingleOpResolver>(type, registration);
         BuildInterpreter({GetShape(input_)}, /*num_threads=*/-1,
                          /*allow_fp32_relax_to_fp16=*/false,
                          /*apply_delegate=*/false, /*allocate_and_delegate=*/true);
       }
  
       void SetInput(const std::vector<float>& data) {
         PopulateTensor(input_, data);
       }
       std::vector<float> GetOutput() { return ExtractVector<float>(output_); }
  
      protected:
       int input_;
       int output_;
     };
  
    ```

Update both tests (SigmoidFloat32ExtremeInput and TanhFloat32ExtremeInput) to use NonDelegatedFloatActivationsOpModel instead of FloatActivationsOpModel<float>.

For example, update SigmoidFloat32ExtremeInput to:

```cpp
  TEST_P(LogisticOpTest, SigmoidFloat32ExtremeInput) {
    NonDelegatedFloatActivationsOpModel m(
        GetRegistration(), BuiltinOperator_LOGISTIC,
        /*input=*/{TensorType_FLOAT32, {1, 6}});
    m.SetInput({
        5e29f,
        1e25f,
        -5e29f,
        -1e25f,
        std::numeric_limits<float>::max(),
        -std::numeric_limits<float>::max(),
    });
    ASSERT_EQ(m.Invoke(), kTfLiteOk);
    EXPECT_THAT(m.GetOutput(), ElementsAreArray(ArrayFloatNear({
                                   1.0f,
                                   1.0f,
                                   0.0f,
                                   0.0f,
                                   1.0f,
                                   0.0f,
                               })));
  }
```
This will ensure the tests pass under all build configurations (including testing defaults that enable XNNPACK). Once you push this update, we should be good to merge!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @Kinetic-Labs-GT,

   get well soon !
  
  Just reiterating the remaining items. Yes, we absolutely want to preserve `NaN` propagation. Standard IEEE-754 rules should apply, meaning that `Sigmoid(NaN)` and `Tanh(NaN)` must evaluate to `NaN` rather than saturating to a default float limit.
  
  We can handle this elegantly by using a simple ternary check before clamping, which avoids complex branching and preserves the `NaN` value:
  
  ```cpp
  const float clamped_val = std::isnan(val) ? val : std::max(-100.0f, std::min(100.0f, val)); // For Logistic/Sigmoid
  
  ```

```cpp
  const float clamped_val = std::isnan(val) ? val : std::max(-20.0f, std::min(20.0f, val)); 
```
// For Tanh
Please update both tensorflow/lite/kernels/internal/reference/logistic.h and tensorflow/lite/kernels/internal/reference/tanh.h with this approach.

In addition, please integrate the NonDelegatedFloatActivationsOpModel helper class we requested in our previous message to bypass default (XNNPACK) delegation during test runs. You can also append a NaN check inside the unit tests to verify the propagation:

```cpp
  TEST_P(LogisticOpTest, SigmoidFloat32ExtremeInput) {
    NonDelegatedFloatActivationsOpModel m(
        GetRegistration(), BuiltinOperator_LOGISTIC,
        /*input=*/{TensorType_FLOAT32, {1, 7}});
    m.SetInput({
        5e29f,
        1e25f,
        -5e29f,
        -1e25f,
        std::numeric_limits<float>::max(),
        -std::numeric_limits<float>::max(),
        std::numeric_limits<float>::quiet_NaN(),
    });
    ASSERT_EQ(m.Invoke(), kTfLiteOk);
    auto output = m.GetOutput();
    EXPECT_THAT(std::vector<float>(output.begin(), output.begin() + 6),
                ElementsAreArray(ArrayFloatNear({
                                   1.0f,
                                   1.0f,
                                   0.0f,
                                   0.0f,
                                   1.0f,
                                   0.0f,
                               })));
    EXPECT_TRUE(std::isnan(output[6]));
  }
```
Please apply these updates and push them to your branch, and we'll be ready to finalize the merge!

**Comment by Kinetic-Labs-GT** [comment]
Hi @dmiltr3,

I have updated the PR to incorporate your feedback regarding IEEE-754 NaN propagation and test delegation:

1. **NaN-Preserving Ternary Clamping:** Updated both `reference/logistic.h` and `reference/tanh.h` to use the non-branching ternary checks (`std::isnan(val) ? val : ...`). This successfully keeps `NaN` propagation intact while applying saturation protection for infinite/extreme boundaries.
2. **Bypassing Delegate Optimization:** Integrated the `NonDelegatedFloatActivationsOpModel` helper into `activations_test.cc`. This explicitly prevents default XNNPACK delegation for our extreme testing paths, isolating our evaluation checks strictly to the reference kernels.
3. **Expanded Test Vectors:** Updated both `SigmoidFloat32ExtremeInput` and `TanhFloat32ExtremeInput` to support an updated shape of `{1, 7}`, introducing a `quiet_NaN()` element alongside a trailing assertion (`EXPECT_TRUE(std::isnan(output[6]));`) to ensure clean propagation.
4. **Git Squashing:** Cleaned up the entire development history by squashing all updates into a single commit on top of the latest upstream master.

The local Bazel validation target (`//tensorflow/lite/kernels:activations_test`) passes cleanly. The PR is ready for your internal testing suite and final merging!

**Review by dmiltr3 — APPROVED** [review]

**Comment by dmiltr3** [comment]
Unfortunately, we have to decline this PR. This code will introduce numerical behavior changes to TFLite. We would like avoid having it unless absolutely necessary.  The reference kernels are meant to define the operator specs in a simple and executable way.  Production use cases usually go to LiteRT+XNNPack. Please check how they are implemented in XNNPack and make changes there instead.
